c2rust-refactor/src/scripting/ast_visitor.rs

```diff
@@ -0,0 +1,493 @@
+                        BindingMode::ByValue(Immutable) => "ByValueImmutable",
+                        BindingMode::ByValue(Mutable) => "ByValueMutable",
+                    })?;
+                    ast.set("ident", ident.as_str().get())?;
+                },
+                _ => return Err(LuaError::external(format!("unimplemented pattern type: {:?}", pat.node))),
+            }
+
+            Ok(ast)
+        })
+    }
+}
+
+impl<'lua> IntoLuaAst<'lua> for Crate {
+    fn into_lua_ast(self, ctx: &TransformCtxt, lua_ctx: LuaContext<'lua>) -> LuaResult<LuaTable<'lua>> {
+        let ast = lua_ctx.create_table()?;
+
+        ast.set("type", "Crate")?;
+        ast.set("module", self.module.into_lua_ast(ctx, lua_ctx)?)?;
+
+        Ok(ast)
+    }
+}
+
+impl<'lua> IntoLuaAst<'lua> for Mod {
+    fn into_lua_ast(self, ctx: &TransformCtxt, lua_ctx: LuaContext<'lua>) -> LuaResult<LuaTable<'lua>> {
+        let ast = lua_ctx.create_table()?;
+        let items = self.items
+            .into_iter()
+            .map(|item| item.into_lua_ast(ctx, lua_ctx))
+            .collect::<LuaResult<Vec<_>>>();
+
+        ast.set("type", "Mod")?;
+        ast.set("inline", self.inline)?;
+        ast.set("items", iter_to_lua_array(items?.into_iter(), lua_ctx)?)?;
+
+        Ok(ast)
+    }
+}
+
+impl<'lua> IntoLuaAst<'lua> for P<Item> {
+    fn into_lua_ast(self, ctx: &TransformCtxt, lua_ctx: LuaContext<'lua>) -> LuaResult<LuaTable<'lua>> {
+        let ast = lua_ctx.create_table()?;
+
+        ast.set("type", "Item")?;
+
+        self.and_then(|item| {
+            ast.set("id", item.id.as_u32())?;
+            ast.set("ident", item.ident.name.as_str().get())?;
+
+            match item.node {
+                ItemKind::ExternCrate(opt_name) => {
+                    ast.set("kind", "ExternCrate")?;
+
+                    if let Some(name) = opt_name {
+                        ast.set("name", name.as_str().get())?;
+                    }
+                },
+                ItemKind::Mod(module) => {
+                    ast.set("kind", "Mod")?;
+                    ast.set("mod", module.into_lua_ast(ctx, lua_ctx)?)?;
+                },
+                ItemKind::Use(_use_tree) => {
+                    ast.set("kind", "Use")?;
+                    // ast.set("use_tree", use_tree.into_lua_ast(ctx, lua_ctx))?;
+                },
+                ItemKind::Fn(decl, _header, _generics, block) => {
+                    ast.set("kind", "Fn")?;
+                    ast.set("decl", decl.into_lua_ast(ctx, lua_ctx)?)?;
+                    ast.set("block", block.into_lua_ast(ctx, lua_ctx)?)?;
+                },
+                ItemKind::Struct(_variant_data, _generics) => {
+                    ast.set("kind", "Struct")?;
+                },
+                ItemKind::Impl(..) => {
+                    ast.set("kind", "Impl")?;
+                }
+                ref e => warn!("Item IntoLuaAst kind: {:?}", e),
```

Comment:
Maybe use `unimplemented!` here so we get an error and we know why.

Reply:
I did this originally but it became annoying because it would often panic on unimplemented nodes there weren't even relevant to what I was scripting